bookstore/clone.py

```diff
@@ -0,0 +1,112 @@
+
+
+BOOKSTORE_FILE_LOADER = FileSystemLoader(PACKAGE_DIR)
+
+
+class BookstoreCloneHandler(IPythonHandler):
+    """Clone a notebook from s3."""
+
+    def initialize(self):
+        # create an easy helper to get at our bookstore settings quickly
+        self.bookstore_settings = BookstoreSettings(config=self.config)
+
+        self.session = aiobotocore.get_session()
+
+    async def _clone(self, s3_bucket, file_key):
+        path = file_key
+
+        self.log.info(f"bucket: {s3_bucket}")
+        self.log.info(f"key: {file_key}")
+        full_s3_path = s3_path(s3_bucket, path)
+
+        async with self.session.create_client(
+            's3',
+            aws_secret_access_key=self.bookstore_settings.s3_secret_access_key,
+            aws_access_key_id=self.bookstore_settings.s3_access_key_id,
+            endpoint_url=self.bookstore_settings.s3_endpoint_url,
+            region_name=self.bookstore_settings.s3_region_name,
+        ) as client:
+            self.log.info("Processing published write of %s", path)
+            obj = await client.get_object(Bucket=s3_bucket, Key=file_key)
+            content = await obj['Body'].read()
+            self.log.info("Done with published write of %s", path)
+
+        resp_content = {"s3_path": full_s3_path}
+        model = {
+            "type": "file",
+            "format": "text",
+            "mimetype": "text/plain",
+            "content": content.decode('utf-8'),
+        }
+        self.set_status(201)
+
+        if 'VersionId' in obj:
+            resp_content["versionID"] = obj['VersionId']
+        return model
+
+    @web.authenticated
+    async def get(self):
+        """Open a page that will allow you to clone a notebook from a specific bucket.
+        """
+        s3_bucket = self.get_argument("s3_bucket", "")
+        # s3_paths module has an s3_key function; file_key avoids confusion
+        file_key = self.get_argument("s3_key", "")
+        self.log.info("About to clone from %s", file_key)
+
+        if file_key == '' or file_key == '/':
+            raise web.HTTPError(400, "Must have a key to clone from")
+        self.set_header('Content-Type', 'text/html')
+        base_uri = f"{self.request.protocol}://{self.request.host}"
+        template_params = {
+            "s3_bucket": s3_bucket,
+            "s3_key": file_key,
+            "base_uri": base_uri,
+            "default_url_base": self.default_url,
+        }
+
+        self.write(self.render_template('clone.html', **template_params))
+
+    @web.authenticated
+    async def post(self):
+        """Clone a notebook to a given path.
+
+        The payload type for the request should be 
+        {
+          "s3_bucket": string,
+          "s3_key": string,
+          "target_path"?: string
+        }
+        The response payload should match the standard Jupyter contents API POST response.
+        """
+        model = self.get_json_body()
+        s3_bucket = model.get("s3_bucket", "")
```

Comment:
This seems wrong, an empty string bucket if the key isn't found would blow up later in the clone call. Shouldn't it fail fast here?